rfcs/20190116-embedding-partitioned-variable.md

```diff
@@ -0,0 +1,498 @@
+
+
+### Partitioned Layer
+
+
+A layer of a network can sometimes be partitioned. We call it partitioned layer in this document. This is another use case of `PartitionedVariable`.
+
+The main reason people partition a layer is to balance parameter servers' load.
+
+Some layers can be huge. For example the logits layer can be extremely large when there are large number of classes.
+
+Even for mid-sized layer, people want to partition them to have more even distribution of parameter servers' load because people may get bad luck when variables are assigned to parameter servers in a round-robin fashion. See this [issue](https://github.com/tensorflow/tensorflow/issues/24953).
+
+Partitioned layer is only supported by lower level APIs with or without `ParameterServerStrategy` in 1.x.
+
+
+### Current Usage of PartitionedVariable
+
+
+With TF 1.x API users can create a `PartitionedVariable` via `tf.get_variable` API:
+
+```python
+partitioned_var = tf.get_variable(
+    "embedding",
+    shape=(10000, 128),
+    dtype=tf.float32,
+    partitioner=tf.min_max_variable_partitioner(
+        max_partitions=num_ps_replicas, axis=0))
+```
+
+The `partitioned_var` here is an instance of `PartitionedVariable`.
+
+Alternatively users can use `tf.variable_scope` to apply `partitioner` to all variables created under its scope:
+
+```python
+with tf.variable_scope(partitioner=tf.min_max_variable_partitioner(
+        max_partitions=num_ps_replicas, axis=0)):
+  tf.keras.layers.Embedding(...)
+```
+
+Both these cases will also work when they are under a `ParameterServerStrategy`'s scope. However `MirroredStrategy` and `CollectiveAllReduceStrategy` will [error out](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/distribute/distribute_lib.py#L872) if a `PartitionedVariable` is about to be created.
+
+These APIs `tf.get_variable` and `tf.variable_scope` will be gone in TF 2.0.
+
+
+### Possible Problems with The `PartitionedVariable` Class 
+
+
+Although we have a class called `PartitionedVariable`, we often concatenate all of their components to pretend it is a single variable. With the concatenation, the `partitioned_var` would only serve for loadbalancing purposes. For example, `tf.matmul(partitioned_var, inputs)` would concatenate the list of variables in the `partitioned_var` before the execution of `matmul` as if the `partitioned_var` is a single variable. 
+
+However, to parallelize computation, people have to write methods that respect the partitions, e.g. `tf.nn.sampled_softmax_loss`. Therefore, this class has been overloaded by different use cases.
+
+On the other hand, when `PartitionedVariable` is used for sharded embeddings, `partitioned_strategy` has to be kept consistent when it is required by several methods down the stream such as `tf.nn.embedding_lookup` and `tf.nn.nce_loss`. This is uncessary for users and any inconsistency would lead to subtle bugs. 
```

Comment:
Maybe we could elaborate more about the justification.
```suggestion
On the other hand, when `PartitionedVariable` is used for sharded embeddings, `partitioned_strategy` has to be kept consistent when it is required by several methods down the stream such as `tf.nn.embedding_lookup` and `tf.nn.nce_loss` which requires an additional `partition_strategy` argument. This is uncessary for users and any inconsistency would lead to subtle bugs. 
```